drivers: spi: stm32H7 SPI driver with DMA

FRASTM opened the pull request:
Adaptation of the spi driver for the stm32H7 serie when DMA is used for data transfers.

It requires additional LL functions to read the address of the stm32H7  SPI Tx and SPI Rx registers for the DMA
- LL_SPI_DMA_GetRxRegAddr
- LL_SPI_DMA_GetTxRegAddr
--> see PR https://github.com/zephyrproject-rtos/hal_stm32/pull/128

The control of the SPI config and status register differs from other stm32 mcu.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: spi: stm32h7 spi drivers with DMA transfer

This patch controls the SPI of the stm32H7 mcu when using
the DMA transfer. Starting and ending the transfer
are specific according to the RefMan.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
tests: drivers: spi loopback with DMA testing on stm32h743 nucleo

with this config, the testcase PASSED on the nucleo_h743zi
using DMA for SPI transfer (through DMAMUX request)
Connect pins D11 and D12 on the ARDUIno connector CN7

Signed-off-by: Francois Ramu <francois.ramu@st.com>

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -722,9 +730,18 @@ static int transceive_dma(const struct device *dev,
 
 	/* This is turned off in spi_stm32_complete(). */
 	spi_stm32_cs_control(dev, true);
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
```
Break a line before this

FRASTM replied:
done

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -722,9 +730,18 @@ static int transceive_dma(const struct device *dev,
+	}
+#else
+	LL_SPI_Enable(spi);
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
```
Break a line after this section

FRASTM replied:
done

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -743,9 +760,11 @@ static int transceive_dma(const struct device *dev,
+#if !defined(CONFIG_SOC_SERIES_STM32H7X)
 		LL_SPI_EnableDMAReq_RX(spi);
 		LL_SPI_EnableDMAReq_TX(spi);
 
+#endif /* ! CONFIG_SOC_SERIES_STM32H7X */
```
Move this line before the blank line.

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -764,9 +783,11 @@ static int transceive_dma(const struct device *dev,
 		LL_SPI_DisableDMAReq_TX(spi);
 		LL_SPI_DisableDMAReq_RX(spi);
 
+#endif /* ! CONFIG_SOC_SERIES_STM32H7X */
```
Move this line before the blank line.

FRASTM replied:
done

erwango commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -56,9 +56,10 @@ struct spi_stm32_data {
 	defined(CONFIG_SOC_SERIES_STM32U5X)
 	return LL_SPI_IsActiveFlag_TXP(spi);
+#elif defined(CONFIG_SOC_SERIES_STM32H7X)
+	return (LL_SPI_IsActiveFlag_TXC(spi));
```
Remove superfluous "()"

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -121,7 +121,12 @@ static int spi_stm32_dma_tx_load(const struct device *dev, const uint8_t *buf,
 		}
 	}
 
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	/* use direct register location until the LL_SPI_DMA_GetTxRegAddr exists */
+	blk_cfg->dest_address = (uint32_t)&(cfg->spi->TXDR);
+#else
 	blk_cfg->dest_address = (uint32_t)LL_SPI_DMA_GetRegAddr(cfg->spi);
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
```
Similarly to what is done in the rest of the driver to ease readability let's use an abstraction function (for instance `ll_func_get_tx_reg_addr`).

FRASTM replied:
using` ll_func_get_reg_addr(spi, SPI_STM32_DMA_TX);`

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -180,7 +185,12 @@ static int spi_stm32_dma_rx_load(const struct device *dev, uint8_t *buf,
 		}
 	}
 
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	/* use direct register location until the LL_SPI_DMA_GetRxRegAddr exists */
+	blk_cfg->source_address = (uint32_t)&(cfg->spi->RXDR);
+#else
 	blk_cfg->source_address = (uint32_t)LL_SPI_DMA_GetRegAddr(cfg->spi);
+#endif /* CONFIG_SOC_SERIES_STM32H7X */
```
Let's use an abstraction function

FRASTM replied:
using ll_func_get_reg_addr(spi, SPI_STM32_DMA_RX);

FRASTM commented:
Add the testcase tests/drivers/spi/spi_loopback with DMA for the stm32h743zi nucleo board :
west build -p auto -b nucleo_h743zi tests/drivers/spi/spi_loopback/

Requires #41560

--> connect pin (D11) SPI 1 MOSI to (D12) SPI1 MISO on the target board to pass the test.

FRASTM commented:
adding generic function `ll_func_get_reg_addr` in the spi_ll_stm32.h  to retreive the SPi data register address for the DMA.

ABOSTM commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -53,12 +56,30 @@ struct spi_stm32_data {
-	defined(CONFIG_SOC_SERIES_STM32H7X) || \
 	defined(CONFIG_SOC_SERIES_STM32U5X)
 	return LL_SPI_IsActiveFlag_TXP(spi);
+#elif defined(CONFIG_SOC_SERIES_STM32H7X)
```
As per our discussion, it was good to check TXP because space is available in FIFO, at least for Interrupt and Polling use cases. But it maybe necessary to differentiate DMA use case where it is necessary to check transfert complete after DMA is complete (DMA is complete means SPI is filled with DMA but there may be remaininf data in FIFO not yet send on bus)

FRASTM replied:
Yes. In case of DMA transfer, after  the DMA Tx is done it is necessary to check the SPI Tx before filling the next DMA packets or ending the transfer. Depending on the soc, this is possible by
- checking the LL_SPI_IsActiveFlag_TXC for stm32H7 or stm32U5
- a combination of LL_SPI_IsActiveFlag_TXE  and LL_SPI_IsActiveFlag_BSY

FRASTM replied:
The stm32H7 and stm32U5 (and stm32MP1)  have the same SPI IP.  The TXC flag is tested ito report the tx fifo empty status  as follows:
```
static inline uint32_t ll_func_tx_is_empty(SPI_TypeDef *spi)
{
#if defined(CONFIG_SOC_SERIES_STM32MP1X) || \
	defined(CONFIG_SOC_SERIES_STM32U5X) || \
	defined(CONFIG_SOC_SERIES_STM32H7X)
	return LL_SPI_IsActiveFlag_TXC(spi);
#else
	return LL_SPI_IsActiveFlag_TXE(spi);
#endif
}

```

FRASTM replied:
commit 14fa24c944c29cb00078c7aa1fa43b61979ef7fd

FRASTM committed:
drivers: spi: stm32 ll driver disable the instance after completion

The spi_stm32_complete() is checking spi flags which are valid when the
intance is still enable: disable afterwards. No more need to disable
the DMA transfer then

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM committed:
drivers: spi: stm32 driver with dma checking spi busy status flag

Defines a function to control the spi busy state during DMA transfer.
After the DMA Tx, the spi might still have some data to Transmit.
The driver must wait for the SPI Tx before sending the next packet.
This is not required for the Rx part as the DMA Rx is already done.
Some mcus like stm32H7 have a TX complete bit, other must wait for the
TXE and BSY line.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented:
1. Rebase
2. Add a dedicated function for DMA transfer to control that SPI has no more Tx in progress: ll_func_spi_dma_tx(). It depends on the stm32mcu to check the pertinent SPI flag(s).
3. The SPI  is now disabled after the spi_stm32_complete() to avoid reading inconsistent SPI flags afterwards. Moreover there is no need to disable the DMA request, as the SPI instance will be totally off.
4.

FRASTM commented:
tests/drivers/spi/spi_loopback running :
- on the nucleo_g474re, nucleo_ h743zi in both Interrupt or DMA modes
- on the stm32g071rb in Interrupt mode

erwango commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -53,6 +53,18 @@ struct spi_stm32_data {
 };
 
+#ifdef CONFIG_SPI_STM32_DMA
+static inline uint32_t ll_func_spi_dma_tx(SPI_TypeDef *spi)
```
Function name is not specific enough. Can you come up with a more meaningful name? (busy, empty, ...)

FRASTM replied:
renaming to "ll_func_spi_dma_busy"

erwango commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -53,6 +53,18 @@ struct spi_stm32_data {
+	return LL_SPI_IsActiveFlag_TXC(spi);
+#else
+	return (LL_SPI_IsActiveFlag_TXE(spi) &&
+	!LL_SPI_IsActiveFlag_BSY(spi));
```
indentation

FRASTM replied:
indented

erwango commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -54,6 +57,21 @@ struct spi_stm32_data {
 };
 
 #ifdef CONFIG_SPI_STM32_DMA
+static inline uint32_t ll_func_get_reg_addr(SPI_TypeDef *spi, uint32_t location)
```
Let's be a bit more specific: `ll_func_dma_get_reg_addr`

FRASTM replied:
renaming to "ll_func_dma_get_reg_addr"

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -723,7 +723,18 @@ static int transceive_dma(const struct device *dev,
 	spi_stm32_cs_control(dev, true);
 
+#if defined(CONFIG_SOC_SERIES_STM32H7X)
+	/* SPI CFG1 register is write protected when SPI is enabled */
```
We don't see direct link with the code and the comment, can you revisit ?

FRASTM replied:
comment changed to "* set request before enabling (else SPI CFG1 reg is write protected)"

FRASTM committed:
drivers: spi: stm32 spi instance of stm32H7 mcu

A generic function is used to give the register address to the DMA.
The SPI of the stm32H7 serie has two data registers for Tx and Rx
When the DMA is getting the address it differs between Rx and Tx.
As the stm32cube/stm32h7xx/drivers/include/stm32h7xx_ll_spi.h
has no such LL functions, the register address is get direclty.

Signed-off-by: Francois Ramu <francois.ramu@st.com>

FRASTM commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -152,10 +173,11 @@ static inline void ll_func_disable_int_errors(SPI_TypeDef *spi)
 	defined(CONFIG_SOC_SERIES_STM32U5X)
 	return (!LL_SPI_IsActiveFlag_MODF(spi) &&
 		!LL_SPI_IsActiveFlag_TXC(spi));
+#elif defined(CONFIG_SOC_SERIES_STM32H7X)
```
This is the busy wait sequence used by the spi_stm32_complete() ; in case of stm32H7, the EOT flag of the SPI status reg.  is monitored

FRASTM replied:
The stm32H7 and stm32U5 (and stm32MP1)  have the same SPI IP.  The TXC flag is tested ito report the tx fifo empty status  as follows:
```
static inline uint32_t ll_func_tx_is_empty(SPI_TypeDef *spi)
{
#if defined(CONFIG_SOC_SERIES_STM32MP1X) || \
	defined(CONFIG_SOC_SERIES_STM32U5X) || \
	defined(CONFIG_SOC_SERIES_STM32H7X)
	return LL_SPI_IsActiveFlag_TXC(spi);
#else
	return LL_SPI_IsActiveFlag_TXE(spi);
#endif
}

```

FRASTM replied:
commit 3a117f7c4df2df618ccf0b57a4c4058ff136d4d8  for the `ll_func_spi_is_busy` function

FRASTM commented:
change function names and comment as per erwango's review

erwango commented on drivers/spi/spi_ll_stm32.c:
```diff
@@ -743,8 +754,11 @@ static int transceive_dma(const struct device *dev,
 			break;
 		}
 
+#if !defined(CONFIG_SOC_SERIES_STM32H7X)
+		/* toggle the DMA request to restart the transfer */
 		LL_SPI_EnableDMAReq_RX(spi);
 		LL_SPI_EnableDMAReq_TX(spi);
+#endif /* ! CONFIG_SOC_SERIES_STM32H7X */
```
Naive question: Can we apply the H7 behavior to all series ?

FRASTM replied:
It seems so. 
The Ref Man of the stm32H7 explicitly requires not to change the DMAReq after the SPI enabled. On other series the driver does not follow the recommended sequence (ex. of stm32G4): 
```
When starting communication using DMA, to prevent DMA channel management raising
error events, these steps must be followed in order:
1. Enable DMA Rx buffer in the RXDMAEN bit in the SPI_CR2 register, if DMA Rx is
used.
2. Enable DMA streams for Tx and Rx in DMA registers, if the streams are used.
3. Enable DMA Tx buffer in the TXDMAEN bit in the SPI_CR2 register, if DMA Tx is used.
4. Enable the SPI by setting the SPE _bit.
```
(We see that step 4 comes before step 1)

There the DMA Tx/Rx request bit toggling is used by the driver to control the DMA transfer (assuming the stm32H7 just relies on the spi_dma_move_buffers to progress in the DMA transfer as DMA Tx/Rx request does not change during the DMA transfer).

erwango reviewed (approved):
Approved, assuming @ABOSTM's requests will be fixed

FRASTM commented:
adding one commit 14fa24c944c29cb00078c7aa1fa43b61979ef7fd for the `ll_func_tx_is_empty` function

ABOSTM commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -152,10 +184,11 @@ static inline void ll_func_disable_int_errors(SPI_TypeDef *spi)
 	return (!LL_SPI_IsActiveFlag_MODF(spi) &&
 		!LL_SPI_IsActiveFlag_TXC(spi));
+#elif defined(CONFIG_SOC_SERIES_STM32H7X)
+	return LL_SPI_IsActiveFlag_EOT(spi);
```
You said IP is the same for H7 and U5, so is it normal to have here a different ?

FRASTM replied:
This change is now tested in SPI with INTerrupt  mode

FRASTM commented:
rebase
+ align the `ll_func_spi_is_busy` function checking the EOT flag for stm32U5 and stm32H7 (and stm32MP1 also) as they have the ame SPI IP

ABOSTM commented on drivers/spi/spi_ll_stm32.h:
```diff
@@ -90,7 +90,7 @@ static inline uint32_t ll_func_tx_is_empty(SPI_TypeDef *spi)
 	defined(CONFIG_SOC_SERIES_STM32H7X) || \
 	defined(CONFIG_SOC_SERIES_STM32U5X)
-	return LL_SPI_IsActiveFlag_TXP(spi);
+	return LL_SPI_IsActiveFlag_TXC(spi);
```
I think this commit should be removed, and we should keep TXP:
TXP mean there is space in the FIFO,
and ll_func_tx_is_empty() is used in spi_stm32_shift_m() spi_stm32_shift_s() 
So as soon as there is space in FIFO we can transfer a new packet in the FIFO.
No need to wait for transfer complete.

FRASTM replied:
Agree, checking the TXP flag instead of TXC (removing the commit)

FRASTM commented:
`ll_func_tx_is_empty` checking the TXP flag instead of TXC (optimizing the SPI  fifo TX control in INT mode)

ABOSTM reviewed (approved):
LGTM

ycsin commented:
I wonder if some of the commits here should be applied to v2.7 LTS? @FRASTM @ABOSTM @erwango

erwango commented:
> I wonder if some of the commits here should be applied to v2.7 LTS? @FRASTM @ABOSTM @erwango

@ycsin Only fixes are subject to backport, so this one doesn't apply.

ycsin commented:
@erwango I'm wondering if 280a61f494bea072d97b49c7b26a6d611f7d17f1 & 824d06b612ae07ea7239d6d4998ba69836da260b are bug fixes?

erwango commented:
> @erwango I'm wondering if [280a61f](https://github.com/zephyrproject-rtos/zephyr/commit/280a61f494bea072d97b49c7b26a6d611f7d17f1) & [824d06b](https://github.com/zephyrproject-rtos/zephyr/commit/824d06b612ae07ea7239d6d4998ba69836da260b) are bug fixes?

To me these are part of enabling SPI dma on H7. Also, I'm not quite in favor of cherry-picking commits among main branch. Two reasons:
- This will become difficult to maintain
- Any one can ask for this or that commit, and we need a clear line to say "yes" or "no", and anyway there will be people not satisfied in the end. So sticking at the PR level seems better to me, as this is an easy rule to understand and apply.
